### ISSUE TITLE
Open recently closed

### ISSUE THREAD

**Issue by guiled** [stoically/firefox-add-on-temporary-containers#28]
- Temporary Containers Version:0.24
- Firefox Version:60
- Operating System + Version:MacOSX High Sierra
- Non-default Options/Preferences in Temporary Containers:
- Other installed Add-ons + Version:

### Actual behavior
Open recently closed tab doesn't work with container tab created by the addon.

### Expected behavior
Closing a container tab created with the addon should add the tab to the browser history

### Steps to reproduce
1. cmd+T Open a new container tab (it is automatically done with the addon)
2. cmd+W Close the tab
3. cmd+shft+T Reopen last closed tab, it reopened a container tab I created manually

### Notes
Not sure it can be fixed...

**Comment by stoically** [comment]
Hi, this is Firefox default behavior and also happens with a fresh profile and without the Temporary Containers Add-on. Nothing I can do about it. If you feel this is a bug, consider filing a Bug over at https://bugzilla.mozilla.org/

**Comment by stoically** [comment]
Ah. this also happens when you actually open a website in a Temporary Container, close it and try to reopen. That's not Firefox default. I'll look into it and see if I can do anything about it. Thanks for reporting.

**Comment by stoically** [comment]
Unfortunately I can't seem to reproduce it anymore, works as expected with these steps.
* Open Tab in Automatic Mode
* Visit example.com
* Close Tab
* Press Ctrl/Cmd+Shift+T

Like I said before - the steps to reproduce which you have provided are Firefox default, there's nothing I can do about it. You'd have to ask Firefox directly to change/fix that.

Thanks for the report and let me know if something else as the steps you provided isn't working as expected.

**Comment by stoically** [comment]
I've encountered this issue again in combination with Multi-Account Containers. Related: https://github.com/mozilla/multi-account-containers/issues/1093 https://github.com/mozilla/multi-account-containers/issues/950

**Comment by stoically** [comment]
This is still more of a problem than it should be. Keeping it open as a reminder. Probably needs support from Firefox platform.

**Comment by stoically** [comment]
This is also reproducible with just permanent containers. Create permanent container, visit example.com, close tab, delete permanent container -> Tab is no longer in "Recently closed tabs" and it's not possible to undo close tab. So this needs to be addressed from Firefox directly. If someone opens a Ticket with Firefox it would be nice to ask for a switch for containers to enable or disable this behavior - because "Deletes History Temporary Containers" probably favour that the closed tabs are not listed in recently closed after the Temporary Container got deleted.

**Comment by SegFault0x0** [comment]
@stoically This bugged me for a while until I looked at your code.  It appears that it's doing exactly what it's supposed/written to do.  In `src/background/container.js` line 305 (`tryToRemove` method), you have:
```javascript
if (tempTabs.length > 0) {
        debug('[tryToRemove] not removing container because it still has tabs', cookieStoreId, tempTabs.length);
        return;
}
```
So as long as I leave *one* tab open for a container, I can close and re-open as many other tabs under that same container, whether they be in the same [sub]domain or opened with ALT+X.  It's not until I close the last one that all of the recent tab history is deleted.  I think this is good, as it works similar to Private Browsing for a specific container (i.e. Closing a container's last tab is like closing a PB window).

However, it'd be nice to have an option (either in Preferences, or perhaps, a popup) that asks whether or not we wish to delete a container when closing its last tab.  If "no", then the container could be deleted whenever the browser window closes.

**Comment by stoically** [comment]
> This bugged me for a while

Me too, actually.

> However, it'd be nice to have an option (either in Preferences, or perhaps, a popup) that asks whether or not we wish to delete a container when closing its last tab. If "no", then the container could be deleted whenever the browser window closes.

That's actually a really good idea to approach this issue - thanks for putting your thoughts down! What I'd say would be viable preferences are:

Remove no longer needed Temporary Containers when
* On browser restart
* After the last tab in it closes
* 10 minutes after the last tab in it closes
* 60 minutes after the last tab in it closes (default)

Remove no longer needed "Deletes History Temporary Containers" when
* On browser restart
* After the last tab in it closes (default)
* 10 minutes after the last tab in it closes
* 60 minutes after the last tab in it closes

I feel like that 60 minutes for normal Temporary Containers is a good timeframe. I don't think I wanted to undo close tab after that time-period anyway and it's still possible to change to "on browser restart". That would make Notifications also more worthwhile (I might even consider enabling them, ha) - because they wont popup all the time when in Automatic Mode.

"Deletes History Temporary Containers" are probably more sensitive and thus the default should be to remove them as soon as possible.

What do you think?

> Unfortunately, this is still intermittent. I can delete all containers, even close the browser, and still re-open tabs. This can happen with or without a Notification displaying.

I'll keep an eye on that when implementing. Thanks for the heads-up.

**Comment by stoically** [comment]
Added delay for removing Temporary Containers. It defaults to "15minutes" for regular containers and "after last tab" for deletes-history containers.

![image](https://user-images.githubusercontent.com/29637501/36445853-c4efdf6e-167f-11e8-8e39-c80906f51c37.png)

About the possibility to being able to still reopen tabs: With "Automatic Mode on Navigation" (no initial reopening) what's happening is that you enter the URL into a non-container tab and on navigation it closes that tab and reopens it in a temporary container. But since you entered something into the address bar, that tab will show up in the "recently closed tabs" as "New Tab". And if you have the preference "Show tabs/windows from last time" in Firefox, recently closed tabs will survive too.

There's no way to remove or prevent "New Tabs" where something got entered into the address bar from showing up in "recently closed tabs" (afaik) with "Automatic Mode in Navigation". With "Reopen Tabs in Automatic Mode" the "New Tabs" don't show up, because nothing gets entered into the address bar.

So if you have the 15minutes delay (default) and do the following within 15minutes in "Automatic Mode on Navigation":
* New tab 1, navigate to example.com
* New tab 2, navigate to example.com
* Close tab 2
* New tab 3, navigate to example.com
* Close tab 3
* Press two times "Ctrl+Shift+T" the second "Undo close tab" will open the "New Tab" with the URL entered. If you keep pressing the shortcut you'll get all actual example.com tabs back though

So yeah, it might be worth considering switching back to the old "Reopen tabs in Automatic Mode" default. (Related: #36)

**Comment by stoically** [comment]
I've published version 0.57beta1 on the beta channel. Any testing would be much appreciated.

**Comment by SegFault0x0** [comment]
>And if you have the preference "Show tabs/windows from last time" in Firefox, recently closed tabs will survive too.
Yup, that's me.

Everything else proposed sounds good!  I'll pull down the changes and test them out when I get a chance tonight.  Great job on the plugin btw, and excellent turnaround time!

**Comment by stoically** [comment]
Just published version 0.58 with the change. Undo close tab is now possible for 15minutes by default. Let me know if something is not working as intended.

**Comment by SegFault0x0** [comment]
Everything seems to be working just fine now.  I can verify that tabs don't delete until 15 min after the last tab in a container is closed.  Good job!
